apps/homescreen/test/unit/app_test.js

```diff
@@ -1,36 +1,564 @@
+    });
+
+    test('should initialise the metadata store', done => {
+      var initStub = sinon.stub(HomeMetadata.prototype, 'init', () => {
+        initStub.restore();
+        done();
+      });
+      new App();
+    });
+
+    test('should initialise the bookmark stores', done => {
+      var initStub = sinon.stub(Datastore.prototype, 'init', () => {
+        initStub.restore();
+        done();
+      });
+      new App();
+    });
+
+    test('should get the list of installed apps', done => {
+      var getAllStub = sinon.stub(MockNavigatormozApps.mgmt, 'getAll', () => {
+        getAllStub.restore();
+        done();
+      });
+      new App();
+    });
+
+    test('should get the list of bookmarked pages', done => {
+      var getAllStub = sinon.stub(Datastore.prototype, 'getAll', () => {
+        getAllStub.restore();
+        done();
+      });
+      new App();
+    });
+
+    test('should remove unknown app entries from metadata store', done => {
+      var removeStub = sinon.stub(HomeMetadata.prototype, 'remove', id => {
+        assert.equal(id, 'def/');
+        removeStub.restore();
+        getStub.restore();
+        getAllStub.restore();
+        done();
+      });
+      var getStub = sinon.stub(HomeMetadata.prototype, 'get', () => {
+        return Promise.resolve([
+          { id: 'abc/', icon: 'abc', order: 0 },
+          { id: 'def/', icon: 'def', order: 1 }
+        ]);
+      });
+      var getAllStub = sinon.stub(Datastore.prototype, 'getAll', () => {
+        return Promise.resolve([
+          { id: 'abc/', data: {} },
+          { id: 'def/', data: {} }
+        ]);
+      });
+      MockNavigatormozApps.mApps = [{
+        id: 0,
+        manifest: {},
+        manifestURL: 'abc',
+        order: 0
+      }];
+
+      app = new App();
+    });
+  });
+
+  suite('App#saveSettings()', () => {
+    test('should restore data', () => {
+      app.small = true;
+      app.saveSettings();
+      app.small = false;
+      app.restoreSettings();
+
+      assert.equal(app.small, true);
+    });
+
+    test('should not restore data if version number differs', () => {
+      app.small = true;
+      app.saveSettings();
+      var tmp = JSON.parse(mockLocalStorage.mRawContent.settings);
+      tmp.version = 'UnknownVersionNumber';
+      mockLocalStorage.mRawContent.settings = JSON.stringify(tmp.settings);
+      app.small = false;
+      app.restoreSettings();
+
+      assert.equal(app.small, false);
+    });
+  });
+
+  suite('App#addApp()', () => {
+    var addAppIconStub;
+
+    setup(() => {
+      addAppIconStub = sinon.stub(app, 'addAppIcon');
+    });
+
+    teardown(() => {
+      addAppIconStub.restore();
+    });
+
+    test('should only accept valid apps', () => {
+      app.addApp({});
+      assert.isFalse(addAppIconStub.called, 'Empty app objects');
+
+      app.addApp({ manifest: { role: 'system' } });
+      assert.isFalse(addAppIconStub.called, 'Excluded role app');
```

Comment:
I don't think this test will pass anymore, we have to add hidden role apps because the role can change in an update - instead, we add it but keep it hidden by setting display:none on the icon container.